src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -230,61 +239,199 @@ CitusGenerateDeferredQueryStrings(CustomScanState *node, EState *estate, int efl
 		RebuildQueryStrings(workerJob->jobQuery, workerJob->taskList);
 
+		EState *executorState = planState->state;
+		ResetExecutionParameters(executorState);
+
 		return;
 	}
 
-	/* at this point, we're about to do the shard pruning */
-	Assert(workerJob->deferredPruning);
+	/* at this point, we're about to do the shard pruning for fast-path queries */
+	Assert(workerJob->deferredPruning && distributedPlan->fastPathRouterPlan);
 	if (jobQuery->commandType == CMD_INSERT)
 	{
 		HandleDeferredShardPruningForInserts(distributedPlan);
+
+		/* TODO: we're assigning it twice for inserts now */
+		workerJob->taskList = FirstReplicaAssignTaskList(workerJob->taskList);
 	}
 	else
 	{
 		HandleDeferredShardPruningForFastPathQueries(distributedPlan);
 	}
+
+
+	if (list_length(distributedPlan->workerJob->taskList) != 1)
+	{
+		/*
+		 * We might have zero shard queries or multi-row INSERTs at this point,
+		 * we only want to cache single task queries.
+		 */
+		return;
+	}
+
+	/*
+	 * As long as the task accesses local node and the query doesn't have
+	 * any volatile functions, we cache the local Postgres plan on the
+	 * shard for re-use.
+	 */
+	Task *task = linitial(distributedPlan->workerJob->taskList);
+	if (TaskAccessesLocalNode(task) &&
+		!contain_volatile_functions(
+			(Node *) originalDistributedPlan->workerJob->jobQuery))
+	{
+		CacheLocalPlanForTask(task, originalDistributedPlan);
+	}
+	else
+	{
+		EState *executorState = planState->state;
+		ResetExecutionParameters(executorState);
+	}
+}
+
+
+/*
+ * ResetExecutionParameters set the parameter list to NULL. See the function
+ * for details.
+ */
+static void
+ResetExecutionParameters(EState *executorState)
+{
+	/*
+	 * We've processed parameters in ExecuteMasterEvaluableFunctions and
+	 * don't need to send their values to workers, since they will be
+	 * represented as constants in the deparsed query. To avoid sending
+	 * parameter values, we set the parameter list to NULL.
+	 */
+	executorState->es_param_list_info = NULL;
+}
+
+
+/*
+ * CacheLocalPlanForTask caches a plan that is local to this node in the
+ * originalDistributedPlan.
+ *
+ * The basic idea is to be able to skip planning on the shards when possible.
+ */
+static void
+CacheLocalPlanForTask(Task *task, DistributedPlan *originalDistributedPlan)
+{
+	PlannedStmt *localPlan = GetCachedLocalPlan(task, originalDistributedPlan);
+	if (localPlan != NULL)
+	{
+		/* we already have a local plan */
+		return;
+	}
+
+	/*
+	 * All memory allocations should happen in the plan's context
+	 * since we'll cache the local plan there.
+	 */
+	MemoryContext oldContext =
+		MemoryContextSwitchTo(GetMemoryChunkContext(originalDistributedPlan));
+
+	/*
+	 * We prefer to use jobQuery (over task->query) because we don't want any
+	 * functions/params have been evaluated in the cached plan.
+	 */
+	Query *shardQuery = copyObject(originalDistributedPlan->workerJob->jobQuery);
+
+	UpdateRelationToShardNames((Node *) shardQuery, task->relationShardList);
```

Comment:
Could we jump straight to shard names here? or is that a separate PR?

Reply:
I couldn't get this question, can you clarify a bit more?

> or is that a separate PR?

This is the last PR :)

Reply:
This function goes through a lot of trouble to replace table RTE with a function RTE. We could instead have another function that simply updates the table RTE to the right OID, since the shard is local.